Backport of:

From 92e5aa3ac6dea431d75c793db4483538ef3cf44a Mon Sep 17 00:00:00 2001
From: cristy <urban-warrior@git.imagemagick.org>
Date: Tue, 13 Jan 2015 12:16:28 +0000
Subject: [PATCH] Avoid a SIGABRT in sun file handling

This is a partial fix for malformed sun file

Bug: http://www.imagemagick.org/discourse-server/viewtopic.php?f=3&t=26857
Bug-debian: https://bugs.debian.org/832465
Origin: upstream, https://github.com/ImageMagick/ImageMagick/commit/6b4aff0f117b978502ee5bcd6e753c17aec5a961
---
 coders/sun.c | 12 +++++++++---
 1 file changed, 9 insertions(+), 3 deletions(-)

Index: imagemagick-6.6.9.7/coders/sun.c
===================================================================
--- imagemagick-6.6.9.7.orig/coders/sun.c	2016-11-15 15:13:40.503482844 -0500
+++ imagemagick-6.6.9.7/coders/sun.c	2016-11-15 15:13:40.499482795 -0500
@@ -403,7 +403,7 @@
       }
     if ((sun_info.length*sizeof(*sun_data))/sizeof(*sun_data) !=
         sun_info.length || !sun_info.length)
-      ThrowReaderException(ResourceLimitError,"MemoryAllocationFailed");
+      ThrowReaderException(ResourceLimitError,"ImproperImageHeader");
     number_pixels=(MagickSizeType) image->columns*image->rows;
     if ((number_pixels*sun_info.depth) > (8*sun_info.length))
       ThrowReaderException(CorruptImageError,"ImproperImageHeader");
@@ -418,11 +418,11 @@
     height=sun_info.height;
     if ((height == 0) || (sun_info.width == 0) || (sun_info.depth == 0) ||
         ((bytes_per_line/sun_info.depth) != sun_info.width))
-      ThrowReaderException(ResourceLimitError,"MemoryAllocationFailed");
+      ThrowReaderException(ResourceLimitError,"ImproperImageHeader");
     bytes_per_line+=15;
     bytes_per_line<<=1;
     if ((bytes_per_line >> 1) != (sun_info.width*sun_info.depth+15))
-      ThrowReaderException(ResourceLimitError,"MemoryAllocationFailed");
+      ThrowReaderException(ResourceLimitError,"ImproperImageHeader");
     bytes_per_line>>=4;
     sun_pixels=(unsigned char *) AcquireQuantumMemory(height,
       bytes_per_line*sizeof(*sun_pixels));
@@ -431,6 +431,12 @@
     if (sun_info.type == RT_ENCODED)
       (void) DecodeImage(sun_data,sun_info.length,sun_pixels,bytes_per_line*
         height);
+    else
+      {
+        if (sun_info.length > (height*bytes_per_line))
+          ThrowReaderException(ResourceLimitError,"ImproperImageHeader");
+        (void) CopyMagickMemory(sun_pixels,sun_data,sun_info.length);
+      }
     sun_data=(unsigned char *) RelinquishMagickMemory(sun_data);
     /*
       Convert SUN raster image to pixel packets.
